rs/cross-chain/proposal-cli/src/ic_admin/mod.rs

```diff
@@ -0,0 +1,106 @@
+#[cfg(test)]
+mod tests;
+
+use crate::git::CompressedWasmHash;
+use crate::proposal::ProposalTemplate;
+use askama::Template;
+use candid::Principal;
+use clap::Args;
+use std::path::PathBuf;
+
+#[derive(Debug, Clone, Args)]
+pub struct IcAdminArgs {
+    /// Use an HSM to sign calls.
+    #[clap(long)]
+    use_hsm: bool,
+
+    /// The slot related to the HSM key that shall be used.
+    #[clap(
+        long = "slot",
+        help = "Only required if use-hsm is set. Ignored otherwise."
+    )]
+    hsm_slot: Option<String>,
+
+    /// The id of the key on the HSM that shall be used.
+    #[clap(
+        long = "key-id",
+        help = "Only required if use-hsm is set. Ignored otherwise."
+    )]
+    key_id: Option<String>,
+
+    /// The PIN used to unlock the HSM.
+    #[clap(
+        long = "pin",
+        help = "Only required if use-hsm is set. Ignored otherwise."
+    )]
+    pin: Option<String>,
+
+    #[clap(long)]
+    /// The id of the neuron on behalf of which the proposal will be submitted.
+    proposer: u64,
+
+    /// The title of the proposal.
+    #[clap(long)]
+    proposal_title: Option<String>,
+}
+
+impl IcAdminArgs {
+    pub fn command_to_submit_proposal(
+        self,
+        proposal: &ProposalTemplate,
+        generated_files: ProposalFiles,
+    ) -> String {
+        let mode = match proposal {
+            ProposalTemplate::Upgrade(_) => "upgrade",
+            ProposalTemplate::Install(_) => "install",
+        }
+        .to_string();
+        IcAdminTemplate {
+            args: self,
+            mode,
+            canister_id: *proposal.canister_id(),
+            wasm_module_path: generated_files.wasm.to_string_lossy().to_string(),
+            wasm_module_sha256: proposal.compressed_wasm_hash().clone(),
+            arg: generated_files.arg.to_string_lossy().to_string(),
+            summary_file: generated_files.summary.to_string_lossy().to_string(),
+        }
+        .render()
+        .expect("failed to render ic-admin template")
+    }
+}
+
+pub struct ProposalFiles {
+    pub wasm: PathBuf,
+    pub arg: PathBuf,
+    pub summary: PathBuf,
+}
+
+#[derive(Template)]
+#[template(path = "submit_with_ic_admin.shx", escape = "none")]
+// The template uses the extension ".shx" to avoid the automatic linting done by "shfmt" on all ".sh" files.
+// This is necessary because the template contains not-yet-valid shell code, that will only be valid after the template is rendered.
```

Comment:
```suggestion
// This is necessary because the template contains not-yet-valid shell code that will only be valid after the template is rendered.
```